### DISCUSSION TITLE
Speed of Jetbot Nano Teleoperation

### DISCUSSION THREAD

**Discussion by Neophyte628** [NVIDIA-AI-IOT/jetbot#439] [Q&A]
Hi, 

Is there away to control the speed of the Jetbot nano during teleoperation? It moves a little too fast.

Also, how do I assign commands to the other controls on the GamePad

best

**Comment by tokk-nv** [comment]
Hi Neophyte628,

Which "[vehicle](http://jetbot.org/v0.4.3/getting_started.html#step-1-pick-your-vehicle)" or the hardware kit are you using?

If that is Waveshare JetBot kit, discussions like below may help you.
- _Suggested Speed Values for Waveshare Jetbot kit_ https://github.com/NVIDIA-AI-IOT/jetbot/discussions/353
- _Slowing down the waveshare Jetbot_  https://github.com/NVIDIA-AI-IOT/jetbot/discussions/377


For assigning functions on other buttons on your gamepad, you can refer to [the notebook](https://github.com/NVIDIA-AI-IOT/jetbot/blob/master/notebooks/teleoperation/teleoperation.ipynb
) and check out the line below.

```
def save_snapshot(change):
    # save snapshot when button is pressed down
    ...

controller.buttons[5].observe(save_snapshot, names='value')
```

**Reply by Neophyte628** [reply]
I can change the speed through software, but i do not know how to modify the software for the gamepad to have motors run at slower speed, as in the software below

left_link = traitlets.dlink((controller.axes[1], 'value'), (robot.left_motor, 'value'), transform=lambda x: -x)
right_link = traitlets.dlink((controller.axes[3], 'value'), (robot.right_motor, 'value'), transform=lambda x: -x)

**Reply by Neophyte628** [reply]
I think I answered my question -- the following enables the jetbot to run at a controllable speed.



#Create buttons to slow/speed-up jetbot when in teleoperation mode

theScale=.5

def callSlowDown(change):
    global theScale
    if change['new']==1:
        theScale=0.8*theScale
        #print(theScale)

def callSpeedUp(change):
    global theScale
    if change['new']==1:
        theScale=1.25*theScale
        #print(theScale)

controller.buttons[0].observe(callSpeedUp,names='value')         
controller.buttons[2].observe(callSlowDown,names='value')    
   
#attach controls to robot

robot = Robot()

left_link = traitlets.dlink((controller.axes[1], 'value'), (robot.left_motor, 'value'), transform=lambda x: -theScale*x)
right_link = traitlets.dlink((controller.axes[5], 'value'), (robot.right_motor, 'value'), transform=lambda x: -theScale*x)    
    


#Create buttons to slow/speed-up jetbot when in teleoperation mode

theScale=.5

def callSlowDown(change):
    global theScale
    if change['new']==1:
        theScale=0.8*theScale
        #print(theScale)

def callSpeedUp(change):
    global theScale
    if change['new']==1:
        theScale=1.25*theScale
        #print(theScale)

controller.buttons[0].observe(callSpeedUp,names='value')         
controller.buttons[2].observe(callSlowDown,names='value')    
   
#attach controls to robot

robot = Robot()

left_link = traitlets.dlink((controller.axes[1], 'value'), (robot.left_motor, 'value'), transform=lambda x: -theScale*x)
right_link = traitlets.dlink((controller.axes[5], 'value'), (robot.right_motor, 'value'), transform=lambda x: -theScale*x)

**Reply by Neophyte628** [reply]
left_link = traitlets.dlink((controller.axes[1], 'value'), (robot.left_motor, 'value'), transform=lambda x: -theScale*x)
right_link = traitlets.dlink((controller.axes[5], 'value'), (robot.right_motor, 'value'), transform=lambda x: -theScale*x)

**Reply by Neophyte628** [reply]
theScale * x
